docs/proposals/20200602-machine-deletion-phase-hooks.md

```diff
@@ -0,0 +1,315 @@
+deleting an instance from a cloud provider, and deleting a node object.
+
+### Hook Implementing Controller (HIC)
+The Hook Implementing Controller describes a controller, other than the
+machine-controller, that adds, removes, and/or responds to a particular
+lifecycle hook.  Each lifecycle hook should have a single HIC, but an HIC
+can optionally manage one or more hooks.
+
+## Summary
+
+Defines a set of annotations that can be applied to a machine which affect the
+linear progress of a machine’s lifecycle after a machine has been marked for
+deletion.  These annotations are optional and may be applied during machine
+creation, sometime after machine creation by a user, or sometime after machine
+creation by another controller or application.
+
+## Motivation
+
+Allow custom and 3rd party components to easily interact with a machine or
+related resources while that machine's reconciliation is temporarily paused.
+This pause in reconciliation will allow these custom components to take action
+after a machine has been marked for deletion, but prior to the machine being
+drained and/or associated instance terminated.
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+
+### Non-Goals/Future Work
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases.
+- Create a mechanism to signal what lifecycle point a machine is at currently.
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine-controller.
+- Require anyone use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+
+### User Stories
+#### Story 1
+(pre-terminate) As an operator, I would like to have the ability to perform
+different actions between the time a machine is marked deleted in the api and
+the time the machine is deleted from the cloud.
+
+For example, when replacing a control plane machine, ensure a new control
+plane machine has been successfully created and joined to the cluster before
+removing the instance of the deleted machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.  This will also prevent
+prolonged periods of having one fewer control plane host in the event the
+replacement instance does not come up in a timely manner.
+
+#### Story 2
+(pre-drain) As an operator, I want the ability to utilize my own draining
```

Comment:
is there any example for a scenario using a custom controller for draining? wouldn't this be solved by letting our drain defaults to be tweakable?

Reply:
Well, there are things that we can tune in the drain library today, and there are things we can't.  For instance, let's say you want to drain all pods not in namespace 'critical-stuff' first, and then drain the pods in 'critical-stuff'  There's not really a way to do that today as the drain library doesn't support it.

One scenario I remember in particular was ensuring certain workloads weren't drained, and we waited until completion.  Certain jobs should never be interrupted, and a custom drain controller might take that into account (I think this can be done with PDBs, but this was the ask).  More or less, this was asked for by someone else, and I think there's some utility in the pattern.  I seem to remember having something like this being a compromise for users that already have their own drainer outside of the machine-controller when discussing adding drain to the machine-controller originally.

Reply:
Thanks @michaelgugino for the reply! I'm making a final pass through here to resolve comments that are noncontroversial. I believe we're all in violent agreement, and your WIP PR seems straightforward and without issue. Hoping to be able to get this merged ASAP.